### DISCUSSION TITLE
Guardrails for functions with inputs that have numerical bounds

### DISCUSSION THREAD

**Discussion by kurt-rhee** [pvlib/pvlib-python#1813] [Ideas]
Hello pvlib,

I was wondering if it would be useful to have some warnings in functions which have numerical inputs that have bounds, but don't error out if the boundary conditions are not met.  Happy to contribute, if it would be useful.

For example, pvlib.tracking.singleaxis takes GCR as an input.  In the documentation the function asks for a number from 0 to 1, but any number could theoretically be used as an input.  If a number outside of 0 and 1 is used, then the function gives interesting results.

```
import warnings

def check_bound(value, lower_bound, upper_bound):
    if value < lower_bound or value > upper_bound:
        warnings.warn(f"Value {value} is outside the allowed range [{lower_bound}, {upper_bound}].")
```

**Comment by kandersolar** [comment]
It's an interesting idea that comes up now and again.  See https://github.com/pvlib/pvlib-python/issues/433 for general related discussion and more recently https://github.com/pvlib/pvlib-python/discussions/1804 for a specific instance.

I have the sense that we aren't interested in adding this kind of checking in the functions.  But there is talk (maybe not public, yet) about adding checks along these lines for ModelChain, maybe as an optional post-processing step on the ModelChainResult or something.

**Reply by kurt-rhee** [reply]
Thanks @kandersolar, #433's speed comments from Mark are very compelling.  I like the idea of a separate "checker" where you could maybe hold all of your inputs in a dictionary and check them separately from the actual runtime of the function.

**Comment by adriesse** [comment]
Range checks could be made conditional, like assertions, to manage a potential performance hit.

**Comment by markcampanelli** [comment]
ArjanCodes just posted a video on using Value Objects (VOs) as function/method parameters to address this issue. I use VOs a lot in a functional programming style with self-validated, immutable VOs as inputs. This usually prevents having to do a gazillion type, range, dimension, etc. checks in the function.

Of course, with pvlib this might mean wrapping/extending numpy.ndarray, pandas.Series, and pandas.DataFrame. That said, potentially VO classes could act as the abstraction layer with a well-defined interface to all these "array" implementations and thus open the door to, say, xarray, or even just plain old Python iterables.

https://youtu.be/CWYwz3iV1g0

**Reply by kurt-rhee** [reply]
This is amazing!
